src/components/codecSupportHelper.ts

```diff
@@ -91,119 +295,285 @@ export function getMaxBitrateSupport(): number {
+                newRes.height
+            )
+        ) {
+            // Return early, since it'll be the best we'll find.
+            return newRes;
+        }
     }
 
-    return h264Profiles;
+    return maxRes;
 }
 
 /**
- * Get the highest H.264 level supported by the active Cast device.
- * @param deviceId - Cast device id.
- * @returns The highest supported H.264 level.
+ * Gets the supported profiles for a given video codec.
+ * @param codec - The video codec in question.
+ * @returns An array of the supported profiles.
  */
-export function getH264LevelSupport(deviceId: DeviceIds): number {
-    switch (deviceId) {
-        case DeviceIds.NESTHUBANDMAX:
-        case DeviceIds.GEN1AND2:
-            return 41;
-        case DeviceIds.GEN3:
-        case DeviceIds.ULTRA:
-            return 42;
-        case DeviceIds.CCGTV:
-            return 51;
-    }
-
-    return 0;
+export function getVideoProfileSupport(codec: VideoCodec): string[] {
+    const possibleProfiles = ((): string[] => {
+        switch (codec) {
+            case VideoCodec.H264:
+                return [
+                    'constrained baseline',
+                    'baseline',
+                    'main',
+                    'high',
+                    'high 10'
+                ];
+            case VideoCodec.H265:
+                return ['main', 'main 10', 'high', 'high 10'];
+            case VideoCodec.AV1:
+                return ['main', 'high', 'professional'];
+            case VideoCodec.VP8:
+                return [''];
+            case VideoCodec.VP9:
+                return ['Profile 0', 'Profile 1', 'Profile 2', 'Profile 3'];
+        }
+    })();
+
+    const supportedProfiles = possibleProfiles.filter((profile) => {
+        const codecString = getCodecString(codec, profile);
+        const mimeType = videoCodecToMimeType(codec);
+
+        return castContext.canDisplayType(mimeType, codecString);
+    });
+
+    return supportedProfiles;
 }
 
 /**
- * Get all H.265 profiles supported by the active Cast device.
- * @param deviceId - Cast device id.
- * @returns All supported H.265 profiles.
+ * Gets the highest level supported by the given codec profile.
+ * @param codec - The codec in question.
+ * @param profile - The profile for the codec.
+ * @param bitDepth - The bit depth of the video.
+ * @returns `number` representing the  highest level supported.
  */
-export function getH265ProfileSupport(deviceId: DeviceIds): string {
-    // These are supported by all Cast devices, excluding audio only devices.
-    if (deviceId === DeviceIds.ULTRA || deviceId === DeviceIds.CCGTV) {
-        return 'high|main|baseline|constrained baseline|high 10';
-    }
-
-    return '';
+export function getVideoCodecHighestLevelSupport(
+    codec: VideoCodec,
+    profile?: string,
+    bitDepth?: number
+): number | undefined {
+    const possibleLevels = ((): number[] => {
+        switch (codec) {
+            case VideoCodec.H264:
```

Comment:
Similar comment to [this](https://github.com/jellyfin/jellyfin-chromecast/pull/744/files#r1917165580), would a specific instance of a derived class for each VideoCodec, that exposed the levels supported as a property be clearer?